<a href="https://colab.research.google.com/github/Rajasurya434/Python-Assignments/blob/main/Assignment10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
df=pd.read_excel('/content/Online retail.xlsx',header=None)

In [4]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


In [5]:
df.describe()

0
count      7501
unique     5176
top     cookies
freq        223

In [6]:
df.isnull().sum()

0    0
dtype: int64

In [7]:
df.head()

0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

In [8]:
df=df.drop_duplicates()

In [9]:
df=df.dropna()

In [12]:

df.columns = ['Transaction']


In [13]:
transactions = df['Transaction'].apply(lambda x: x.split(','))

transactions = transactions.apply(lambda x: [item.strip() for item in x])

print(transactions.head())


0    [shrimp, almonds, avocado, vegetables mix, gre...
1                           [burgers, meatballs, eggs]
2                                            [chutney]
3                                    [turkey, avocado]
4    [mineral water, milk, energy bar, whole wheat ...
Name: Transaction, dtype: object


Splits the string into list of items taking ',' to seperate

In [17]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()

te_ary = te.fit(transactions).transform(transactions)

basket = pd.DataFrame(te_ary, columns=te.columns_)

basket.head()


almonds  antioxydant juice  asparagus  avocado  babies food  bacon  \
0     True               True      False     True        False  False   
1    False              False      False    False        False  False   
2    False              False      False    False        False  False   
3    False              False      False     True        False  False   
4    False              False      False    False        False  False   

   barbecue sauce  black tea  blueberries  body spray  ...  turkey  \
0           False      False        False       False  ...   False   
1           False      False        False       False  ...   False   
2           False      False        False       False  ...   False   
3           False      False        False       False  ...    True   
4           False      False        False       False  ...   False   

   vegetables mix  water spray  white wine  whole weat flour  \
0            True        False       False              True   
1           False        False       False             False   
2           False        False       False             False   
3           False        False       False             False   
4           False        False       False             False   

   whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0              False             False   True        False     False  
1              False             False  False        False     False  
2              False             False  False        False     False  
3              False             False  False        False     False  
4              False              True  False        False     False  

[5 rows x 119 columns]

Convert to binary matrix so that it can be used with aprori algorthmn , we converted raw data to binary matrix .

In [18]:
from mlxtend.frequent_patterns import apriori, association_rules


frequent_itemsets = apriori(basket, min_support=0.01, use_colnames=True)


rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)


rules.head()


antecedents      consequents  antecedent support  \
0              (chicken, chocolate)  (mineral water)            0.021252   
1            (chocolate, olive oil)  (mineral water)            0.023570   
2               (eggs, ground beef)  (mineral water)            0.028787   
3  (frozen vegetables, ground beef)  (mineral water)            0.024536   
4  (frozen vegetables, ground beef)      (spaghetti)            0.024536   

   consequent support   support  confidence      lift  leverage  conviction  \
0            0.299845  0.011012    0.518182  1.728163  0.004640    1.453151   
1            0.299845  0.011978    0.508197  1.694862  0.004911    1.423648   
2            0.299845  0.014490    0.503356  1.678717  0.005858    1.409771   
3            0.299845  0.013331    0.543307  1.811957  0.005974    1.533097   
4            0.229521  0.012558    0.511811  2.229911  0.006926    1.578240   

   zhangs_metric  
0       0.430500  
1       0.419878  
2       0.416291  
3       0.459382  
4       0.565425

2.1252% of transactions contain chicken and chocolate

29.9845% of transactions contain mineral water

1.1012% of transactions contain both chicken, chocolate, and mineral water

51.8182% of transactions that contain chicken and chocolate also contain mineral water

Mineral water is 1.728163 times more likely to be bought with chicken and chocolate than expected by chance



In [22]:
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(basket, min_support=0.01, use_colnames=True)

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)

rules = rules[rules['lift'] > 1.0]
print(rules.head())


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


                        antecedents      consequents  antecedent support  \
0              (chicken, chocolate)  (mineral water)            0.021252   
1            (chocolate, olive oil)  (mineral water)            0.023570   
2               (eggs, ground beef)  (mineral water)            0.028787   
3  (frozen vegetables, ground beef)  (mineral water)            0.024536   
4  (frozen vegetables, ground beef)      (spaghetti)            0.024536   

   consequent support   support  confidence      lift  leverage  conviction  \
0            0.299845  0.011012    0.518182  1.728163  0.004640    1.453151   
1            0.299845  0.011978    0.508197  1.694862  0.004911    1.423648   
2            0.299845  0.014490    0.503356  1.678717  0.005858    1.409771   
3            0.299845  0.013331    0.543307  1.811957  0.005974    1.533097   
4            0.229521  0.012558    0.511811  2.229911  0.006926    1.578240   

   zhangs_metric  
0       0.430500  
1       0.419878  
2       0.4

This fliters data with confidence >0.5 and lift > 1